### PR TITLE
WIP: random group elements

### PR DESCRIPTION
**Pull request by fvanmaele**
This is the first part in the assignment for #12. Testing all small permutation groups from order 2 to 30, I have for 50k samples approximate (90% confidence) uniform distribution between all elements.

I'm working on the basic backtrack and hope to have something workable before the 25th.

### CODE REVIEW

**Comment by fvanmaele** [comment]
I'm not sure how to make the recursive algorithm work with the iterator interface. If it were Python, I'd use `yield` instead of pushing to `L`. In Julia, I found [`Channel`](https://docs.julialang.org/en/v1/manual/asynchronous-programming/) but it's unclear how to adapt this to `iterate`. 

There's an old discussion here: https://discourse.julialang.org/t/from-generator-with-channel-to-iterator/10277/4

but it doesn't apply to newer versions of Julia.

**Comment by fvanmaele** [comment]
To answer the remaining questions on the exercise sheet:
> a general backtrack with an oracle to descend to/prune branches. What kind of information should we pass to it (except current node represented by an element g)? (note: there is no absolutely general search, the oracle is very often problem dependent).

=> We should at least pass the partial base images. The stack-based approach allows to traverse the tree in breadth-first ordering. (Demonstration with `println` in `backtrack_stack!`).

> Implement a method that takes G::AbstractPermGroup and g ∈ G and reconstructs a word in the original generators of G which is equal to g.

My `decompose` method first constructs the stabilizer chain of the permutation group, and then uses a modified element check. The group element `g` can be decomposed into transversal elements, which themselves can be decomposed into a word of generators (and their inverses). To achieve the latter, I've used `SchreierTree` to construct the `PermutationGroup`, and implemented a `trace` helper in `orbits.jl`.

Right now I'm trying to get the tests to run. I got stuck last week because of an assertion error in `schreier_sims` (see my last email). However I noticed I didn't pull in the last fixes, so I will test again.

**Comment by fvanmaele** [comment]
I pushed some more changes, but I'm struggling with some parts. Namely:

1. For sending elements `g` to `h`, this is Example II.16 in the notes. I'm however having trouble with more complex examples, I guess I need to revisit the lecture of 31/05...
2. The tests fail with a cryptic error: `ArgumentError: Package Random not found in current path`. I thought this was something weird with my installation, but it fails on the CI as well.
3. My `decompose` implementation still seems to have some bugs.

In other words there's still a lot left to do. I'm not sure if @JannisHeising got any further than me. The repository is still public, I guess until the start of November? 

Unfortunately this month I still have 2 projects I need to finish for convex optimization, but I'll still try to make some progress.

**Comment by kalmarek** [comment]
@fvanmaele Great! I'll take a look at this this weekend.

to make backtrack search compatible with `iterate` I came up with this code
https://gist.github.com/kalmarek/883501c685ecf22f787eec5d173c143c
(have a look at the debug log included there)

To use this write
```julia
julia> bs = BacktrackSearch(
           ['a':'c', 'x':'z'],
           oracle_descend=bs->isempty(bs.nodes) || (last(bs.nodes) ∈ ('a', 'c', 'x', 'z')),
           node_accept=bs->current_depth(bs)==depth(bs)
       );

julia> collect(bs)
4-element Vector{Vector{Char}}:
 ['a', 'x']
 ['a', 'z']
 ['c', 'x']
 ['c', 'z']
```
Note that we had to make `oracle` to _explicitly accept_ the initial (empty) node of our search tree.

The `yield` instruction from python can be "emulated" by setting one flag and a conditional jump. I guess the dislike of `@goto → @label` made this harder for you than it actually was ;) 

# "Benchmark"
An obligatory "benchmark" follows:
```julia
julia> itrs
3-element Vector{StepRange{Char, Int64}}:
 'a':1:'k'
 'l':1:'q'
 's':1:'z'

julia> prod(length, itrs) # the size of the tree
528

julia> bs = BacktrackSearch(
           itrs,
           oracle_descend=bs->isempty(bs.nodes) || (last(bs.nodes) ∈ ('a':2:'k'...,'k':2:'q'..., 'q', 'x', 'z')),
           node_accept=bs->current_depth(bs) == depth(bs)
       );

julia> @time length(collect(bs))
  0.154247 seconds (162.74 k allocations: 8.232 MiB, 99.94% compilation time)
36

julia> @time length(collect(bs))
  0.000120 seconds (75 allocations: 4.391 KiB)
36
```
Not bad -- it seems that we make just 2 allocations per accepted node. I was aiming for 1 though ;), not sure where the other one is coming from but I didn't investigate.
Of course, as you can see you the types of oracles/node_accepts depend heavily on the problem we're solving.

## Paying the compilation price once
The drawback is that every time we define new `oracle_descend`s/`accept_node` we take hit for compilation. We could alleviate this with
```julia
isleaf(bs::BacktrackSearch) = current_depth(bs)==depth(bs)

struct EndsWithAnyOf{T}
    elts::T
end

(ewa::EndsWithAnyOf)(bs::BacktrackSearch) = isempty(bs.nodes) || last(bs.nodes) in ewa.elts
```
The example above would then become
```julia
julia> bs2 = BacktrackSearch(
           itrs,
           oracle_descend=EndsWithAnyOf(('a':2:'k'...,'k':2:'q'..., 'q', 'x', 'z')),
           node_accept=isleaf
       );
```
